src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/RuntimeAsyncRewriter.cs

```diff
@@ -22,30 +24,65 @@ public static BoundStatement Rewrite(
             return node;
         }
 
-        // https://github.com/dotnet/roslyn/issues/79763: struct lifting
         var variablesToHoist = IteratorAndAsyncCaptureWalker.Analyze(compilationState.Compilation, method, node, isRuntimeAsync: true, diagnostics.DiagnosticBag);
+        var hoistedLocals = ArrayBuilder<LocalSymbol>.GetInstance();
+        var factory = new SyntheticBoundNodeFactory(method, node.Syntax, compilationState, diagnostics);
+        var rewriter = new RuntimeAsyncRewriter(factory, variablesToHoist, hoistedLocals);
+        var thisStore = cacheThisIfNeeded(rewriter);
+        var result = (BoundStatement)rewriter.Visit(node);
 
-        if (variablesToHoist.Count > 0)
+        if (thisStore is not null)
         {
-            foreach (var variable in variablesToHoist)
-            {
-                // Method '{0}' uses a feature that is not supported by runtime async currently. Opt the method out of runtime async by attributing it with 'System.Runtime.CompilerServices.RuntimeAsyncMethodGenerationAttribute(false)'.
-                diagnostics.Add(ErrorCode.ERR_UnsupportedFeatureInRuntimeAsync, variable.GetFirstLocation(), method);
-            }
+            result = factory.Block(hoistedLocals.ToImmutableAndFree(),
+                factory.HiddenSequencePoint(),
+                factory.ExpressionStatement(thisStore),
+                result);
+        }
+        else if (hoistedLocals.Count > 0)
+        {
+            result = factory.Block(hoistedLocals.ToImmutableAndFree(), result);
+        }
+        else
+        {
+            hoistedLocals.Free();
         }
 
-        var rewriter = new RuntimeAsyncRewriter(new SyntheticBoundNodeFactory(method, node.Syntax, compilationState, diagnostics));
-        var result = (BoundStatement)rewriter.Visit(node);
         return SpillSequenceSpiller.Rewrite(result, method, compilationState, diagnostics);
+
+        static BoundAssignmentOperator? cacheThisIfNeeded(RuntimeAsyncRewriter rewriter)
+        {
+            Debug.Assert(rewriter._factory.CurrentFunction is not null);
+            var thisParameter = rewriter._factory.CurrentFunction.ThisParameter;
+            if (rewriter._variablesToHoist.Contains(thisParameter))
+            {
+                // This is a struct or a type parameter, and accessed across an await. We need to replace it with a hoisted local to preserve behavior,
```

Comment:
If I understood correctly, "preserve behavior" means "preserve async behavior" since async methods don't side-effect the original struct. Is that what this means? (nit: consider clarifying the behavior being preserved)
Do we have a test illustrating the significance of this hoisting in terms of observed expected output?